estimatefee.go

```diff
@@ -0,0 +1,434 @@
+// happened, but it should be close enough.
+func (ef *feeEstimator) Rollback(block *btcutil.Block) error {
+	ef.Lock()
+	defer ef.Unlock()
+
+	hash := block.Hash()
+
+	// Find this block in the stack of recent registered blocks.
+	var n int
+	for n = 1; n < len(ef.dropped); n++ {
+		if ef.dropped[len(ef.dropped)-n].hash.IsEqual(hash) {
+			break
+		}
+	}
+
+	if n == len(ef.dropped) {
+		return errors.New("No such block was recently registered.")
+	}
+
+	for i := 0; i < n; i++ {
+		err := ef.rollback()
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// rollback rolls back the effect of the last block in the stack
+// of registered blocks.
+func (ef *feeEstimator) rollback() error {
+
+	// The previous sorted list is invalid, so delete it.
+	ef.cached = nil
+
+	// pop the last list of dropped txs from the stack.
+	last := len(ef.dropped) - 1
+	if last == -1 {
+		// Return if we cannot rollback.
+		return errors.New("Max rollbacks reached.")
+	}
+
+	ef.numBlocksRegistered--
+
+	dropped := ef.dropped[last]
+	ef.dropped = ef.dropped[0:last]
+
+	// where we are in each bin as we replace txs?
+	var replacementCounters [estimateFeeBins]int
+
+	// Go through the txs in the dropped box.
+	for _, o := range dropped.transactions {
+		// Which bin was this tx in?
+		blocksToConfirm := o.mined - o.observed - 1
+
+		bin := ef.bin[blocksToConfirm]
+
+		var counter = replacementCounters[blocksToConfirm]
+
+		// Continue to go through that bin where we left off.
+		for {
+			if counter >= len(bin) {
+				panic("Illegal state: cannot rollback dropped transaction!")
+			}
+
+			prev := bin[counter]
+
+			if prev.mined == ef.lastKnownHeight {
+				prev.mined = mempoolHeight
+
+				bin[counter] = o
+
+				counter++
+				break
+			}
+
+			counter++
+		}
+	}
+
+	// Continue going through bins to find other txs to remove
+	// which did not replace any other when they were entered.
+	for i, j := range replacementCounters {
+		for {
+			l := len(ef.bin[i])
+			if j >= l {
+				break
+			}
+
+			prev := ef.bin[i][j]
+
+			if prev.mined == ef.lastKnownHeight {
+				prev.mined = mempoolHeight
+
+				ef.bin[i] = append(ef.bin[i][0:j], ef.bin[i][j+1:l]...)
+
+				continue
+			}
+
+			j++
+		}
+	}
+
+	ef.lastKnownHeight--
+
+	return nil
+}
+
+// estimateFeeSet is a set of txs that can that is sorted
+// by the fee per kb rate.
+type estimateFeeSet struct {
+	feeRate []float64
+	bin     [estimateFeeBins]uint32
+}
+
+func (b *estimateFeeSet) Len() int { return len(b.feeRate) }
+
+func (b *estimateFeeSet) Less(i, j int) bool {
+	return b.feeRate[i] > b.feeRate[j]
+}
+
+func (b *estimateFeeSet) Swap(i, j int) {
+	b.feeRate[i], b.feeRate[j] = b.feeRate[j], b.feeRate[i]
+}
+
+// EstimateFee returns the estimated fee for a transaction
+// to confirm in confirmations blocks from now, given from
+// the data set we have collected.
+func (b *estimateFeeSet) EstimateFee(confirmations int) float64 {
+	if confirmations <= 0 {
+		return math.Inf(1)
```

Comment:
Rather than reporting an estimated fee of `infinity`, I think an error should be returned here instead. 

An alternative is to modify the `confirmations` parameter to be a `uint` rather than an `int`.